pkg/api/metadata.go

```diff
@@ -0,0 +1,197 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package api
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"reflect"
+	"strings"
+	"time"
+)
+
+const (
+	RelationSelf      = "self"
+	RelationNext      = "next"
+	RelationPrev      = "prev"
+	RelationAlternate = "alternate"
+	RelationLabels    = "https://stormforge.io/rel/labels"
+	RelationTrials    = "https://stormforge.io/rel/trials"
+	RelationNextTrial = "https://stormforge.io/rel/next-trial"
+)
+
+// Metadata is used to hold single or multi-value metadata from list responses.
+type Metadata map[string][]string
+
+func (m Metadata) Title() string {
+	return http.Header(m).Get("Title")
+}
+func (m Metadata) SetTitle(value string) {
+	if value != "" {
+		http.Header(m).Set("Title", value)
+	}
+}
+
+func (m Metadata) Location() string {
+	return http.Header(m).Get("Location")
+}
+func (m Metadata) SetLocation(value string) {
+	if value != "" {
+		http.Header(m).Set("Location", value)
+	}
+}
+
+func (m Metadata) LastModified() time.Time {
+	value, _ := http.ParseTime(http.Header(m).Get("Last-Modified"))
+	return value
+}
+func (m Metadata) SetLastModified(t time.Time) {
+	if !t.IsZero() {
+		http.Header(m).Set("Last-Modified", t.UTC().Format(http.TimeFormat))
+	}
+}
+
+func (m Metadata) Link(rel string) string {
+	for _, rh := range http.Header(m).Values("Link") {
+		for _, h := range strings.Split(rh, ",") {
+			r, l := splitLink(h)
+			if strings.EqualFold(rel, r) {
+				return l
+			}
+		}
+	}
+	return ""
+}
+func (m Metadata) SetLink(rel, link string) {
+	if link == "" {
+		return
+	}
+
+	values := http.Header(m).Values("Link")
+	http.Header(m).Del("Link")
+	var added bool
+	for _, rh := range values {
+		for _, h := range strings.Split(rh, ",") {
+			r, l := splitLink(h)
+			if strings.EqualFold(rel, r) {
+				l = link
+				added = true
+			}
+			http.Header(m).Add("Link", fmt.Sprintf(`<%s>;rel=%q`, l, r))
+		}
+	}
+
+	if !added {
+		http.Header(m).Add("Link", fmt.Sprintf(`<%s>;rel=%q`, link, rel))
+	}
+}
+
+func splitLink(value string) (rel, link string) {
+	for _, l := range strings.Split(value, ";") {
+		l = strings.Trim(l, " ")
+		if l == "" {
+			continue
+		}
+
+		if l[0] == '<' && l[len(l)-1] == '>' {
+			link = strings.Trim(l, "<>")
+			continue
+		}
+
+		p := strings.SplitN(l, "=", 2)
+		if len(p) == 2 && strings.ToLower(p[0]) == "rel" {
+			rel = strings.Trim(p[1], "\"")
+			continue
+		}
+	}
+
+	switch strings.ToLower(rel) {
+	case "previous":
+		rel = RelationPrev
+
+	case "https://carbonrelay.com/rel/labels",
+		"https://carbonrelay.com/rel/triallabels":
+		rel = RelationLabels
+
+	case "https://carbonrelay.com/rel/trials":
+		rel = RelationTrials
+
+	case "https://carbonrelay.com/rel/next-trial",
+		"https://carbonrelay.com/rel/nexttrial":
+		rel = RelationNextTrial
+	}
```

Comment:
Should this be s/carbonrelay.com/stormforge.io/?

Do we want to use the Relation constants above for this ( or add in additional ones )?

Reply:
This code is just normalizing the "legacy" values (i.e. previously supported values for the link relations) so consumers can safely use the constants.

Reply:
I separated this out to a `CanonicalLinkRelation` function to make it a little more clear what was happening.